// bot-monitor-sdk/lib/Config.js
/**
 * Copyright (c) 2017 Baidu, Inc. All Rights Reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

/**
 * @file Config class,contains configuration information about data upload,
 * Under normal circumstances, you do not need to modify the configuration
 * items inside.
 * @author liudesheng01<liudesheng01@baidu.com>
 */

'use strict';

/**
 * Configuration information,contains the data reported configuration add sdk info
 **/
class Config {
    constructor() {
        this._uploadUrl = 'dueros-api.baidu.com';
        this._sdkVersion = '1.0.0';
        this._sdkType = 'nodejs';
        this._uploadPort = 443;
        this._uploadPath = '/uploadmonitordata';
    }

    /**
     * get server upload url
     *
     * @return {string}
     **/
    getUploadurl() {
        return this._uploadUrl;
    }

    /**
     * get sdk version
     *
     * @return {string}
     **/
    getSdkVersion() {
        return this._sdkVersion;
    }

    /**
     * get sdk type
     *
     * @return {string}
     **/
    getSdkType() {
        return this._sdkType;
    }

    /**
     * get upload port
     *
     * @return {string}
     **/
    getUploadPort() {
        return this._uploadPort;
    }

    /**
     * get upload path
     *
     * @return {string}
     **/
    getUploadPath() {
        return this._uploadPath;
    }
}

module.exports = Config;
